### PR TITLE
Add spaces_permissions and users_permissions to grant_token

### PR DESCRIPTION
**Pull request by kleewho**
feat: Add support for spaces and users permissions in grant_token

feat: Add user_id and deprecate uuid when creating new pubnub instance

### CODE REVIEW

File: lib/pubnub/validators/grant_token.rb [hunk 1]
```diff
@@ -45,6 +46,19 @@ def validate_permissions!(arg, name)
           ":#{name} has to be kind of Hash for grant token event."
         ) unless arg.is_a?(Hash)
       end
+
+      def validate_objects_entities_separation!
+        entities_set = not(@spaces_permissions.empty?) ||
+          not(@users_permissions.empty?)
```
Comment (parfeon) [inline lib/pubnub/validators/grant_token.rb:52]: Why not like this:
```suggestion
        entities_set = !@spaces_permissions.empty? ||
          !@users_permissions.empty?
```

File: lib/pubnub/validators/grant_token.rb [hunk 2]
```diff
@@ -45,6 +46,19 @@ def validate_permissions!(arg, name)
...
+        objects_set = not(@channels.empty?) ||
+          not(@channel_groups.empty?) ||
+          not(@uuids.empty?)
```
Comment (parfeon) [inline lib/pubnub/validators/grant_token.rb:55]: Why not like this:
```suggestion
        objects_set = !@channels.empty? ||
          !@channel_groups.empty? ||
          !@uuids.empty?
```

File: spec/examples/grant_token_examples_spec.rb [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+    @callback = -> (_envelope) do
+      @fired = true
+    end
+
+    @pubnub = Pubnub.new(
+      publish_key: "pub-a-mock-key",
+      subscribe_key: "sub-a-mock-key",
+      secret_key: "sec-a-mock-key",
+      uuid: "ruby-test-uuid-client-one",
```
Comment (parfeon) [inline spec/examples/grant_token_examples_spec.rb:15]: @kleewho this will conflict after recent merge of `user_id` changes.

**Review by parfeon — CHANGES_REQUESTED** [review]
@kleewho there is very small changes required

**Review by parfeon — APPROVED** [review]
LGTM!

**Comment by kleewho** [comment]
@client-engineering-bot release as v5.1.0

**Pull request merged**